service/controller/app/resource/chart/desired.go

```diff
@@ -111,6 +110,123 @@ func (r *Resource) GetDesiredState(ctx context.Context, obj interface{}) (interf
 	return chartCR, nil
 }
 
+func (r *Resource) pickRepositoryURL(ctx context.Context, cc *controllercontext.Context, cr v1alpha1.App, chartName string) (string, error) {
+	if len(cc.Catalog.Spec.Repositories) == 1 {
+		return cc.Catalog.Spec.Repositories[0].URL, nil
+	}
+
+	var chart v1alpha1.Chart
+	err := cc.Clients.K8s.CtrlClient().Get(
+		ctx,
+		types.NamespacedName{Name: chartName, Namespace: r.chartNamespace},
+		&chart,
+	)
+	if apierrors.IsNotFound(err) || tenant.IsAPINotAvailable(err) {
+		// Repositories is guaranteed by Custom Resource Definition to have at least one entry.
+		return cc.Catalog.Spec.Repositories[0].URL, nil
+	} else if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	// Check currently selected repository
+	repositoryIndex := -1
+	for i, repo := range cc.Catalog.Spec.Repositories {
+		if strings.Contains(chart.Spec.TarballURL, repo.URL) {
+			repositoryIndex = i
+			break
+		}
+	}
+	if repositoryIndex == -1 {
+		// Could not match current tarballURL to any of Catalog's repositories.
+		// Maybe the list was updated. Let's pick any existing repository.
+		r.logger.Debugf(ctx, "could not match tarball URL %q to any of %q Catalog repositories; using default", chart.Spec.TarballURL, cc.Catalog.Name)
+		return cc.Catalog.Spec.Repositories[0].URL, nil
+	}
+
+	if chart.Status.Release.Status == releaseNotInstalledStatus {
```

Comment:
In the latest `chart-opererator` the `not-installed` status is set when:
- failed to pull, not found or timed out
- also when the chart failed to install and the release is not found

The `unknown-error` can be added when the release fails but there is an already deployed release. I am not sure we need this, but maybe we want to retry on those too.

What I was more concerned about is that there is a `FailedMaxAttempts` field stored in the controller context. The limit is currently [set to 5](https://github.com/giantswarm/chart-operator/blob/0ccda9890974f224b0eb032030d1b3887b39db48/pkg/project/project.go#L11-L16). It [checks the last 5 releases](https://github.com/giantswarm/chart-operator/blob/0ccda9890974f224b0eb032030d1b3887b39db48/service/controller/chart/resource/releasemaxhistory/create.go#L130-L143) in the release history. What [happens after this](https://github.com/giantswarm/chart-operator/blob/0ccda9890974f224b0eb032030d1b3887b39db48/service/controller/chart/resource/releasemaxhistory/create.go#L47-L64) is not super clear to me why we do that. Maybe it has no effects on the fallback charts / failing them multiple times even when the list is long and maybe you thought of it as well.

Reply:
See: https://github.com/giantswarm/chart-operator/pull/889

Reply:
I just realized I don't have to check for `MaxFailedAttempts`. There won't be any Helm release to retry nor any history, since this is only for cases when pulling the chart has failed.